src/devices/machine/tc009xlvc.cpp

```diff
@@ -62,138 +29,102 @@ WRITE8_MEMBER(tc0091lvc_device::vregs_w)
 {
 	if((offset & 0xfc) == 0)
 	{
-		bg0_tilemap->mark_all_dirty();
-		bg1_tilemap->mark_all_dirty();
+		m_bg_tilemap[0]->mark_all_dirty();
+		m_bg_tilemap[1]->mark_all_dirty();
 	}
 
 	m_vregs[offset] = data;
 }
 
-READ8_MEMBER(tc0091lvc_device::tc0091lvc_bitmap_r)
-{
-	return m_bitmap_ram[offset];
-}
-
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_bitmap_w)
-{
-	m_bitmap_ram[offset] = data;
-}
-
-
-READ8_MEMBER(tc0091lvc_device::tc0091lvc_pcg1_r)
+WRITE8_MEMBER(tc0091lvc_device::vram_w)
 {
-	return m_pcg1_ram[offset];
-}
+	// TODO : 0x10000-0x13fff Unused?
+	if ((offset & 0xc000) == 0)
+		return;
 
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_pcg1_w)
-{
-	m_pcg1_ram[offset] = data;
-	m_gfxdecode->gfx(m_gfx_index)->mark_dirty((offset+0x4000) / 32);
-	tx_tilemap->mark_all_dirty();
-}
+	if((m_vram[offset] & mem_mask) == (data & mem_mask))
+		return;
 
-READ8_MEMBER(tc0091lvc_device::tc0091lvc_pcg2_r)
-{
-	return m_pcg2_ram[offset];
-}
-
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_pcg2_w)
-{
-	m_pcg2_ram[offset] = data;
-	m_gfxdecode->gfx(m_gfx_index)->mark_dirty((offset+0xc000) / 32);
-	tx_tilemap->mark_all_dirty();
-}
-
-READ8_MEMBER(tc0091lvc_device::tc0091lvc_vram0_r)
-{
-	return m_vram0[offset];
-}
-
-WRITE8_MEMBER(tc0091lvc_device::tc0091lvc_vram0_w)
-{
-	m_vram0[offset] = data;
-	bg0_tilemap->mark_tile_dirty(offset/2);
-	m_gfxdecode->gfx(m_gfx_index)->mark_dirty((offset+0x8000) / 32);
-	tx_tilemap->mark_all_dirty();
-
-}
-
-READ8_MEMBER(tc0091lvc_device::tc0091lvc_vram1_r)
-{
-	return m_vram1[offset];
+	COMBINE_DATA(&m_vram[offset]);
+	gfx(2)->mark_dirty(offset / 32);
+	switch (offset & 0xf000)
+	{
+		case 0x8000:
+		case 0x9000:
+			m_bg_tilemap[(offset >> 12) & 1]->mark_tile_dirty((offset & 0xfff) / 2);
+			break;
+		case 0xa000:
+			m_tx_tilemap->mark_tile_dirty((offset & 0xfff) / 2);
+			break;
```

Comment:
I thought the whole point of templatizing stuff is to avoid calling switch-cases like this.
In a nutshell this is ugly spaghetti code.